### DISCUSSION TITLE
New feature - STEP Import | User Defined Attributes

### DISCUSSION THREAD

**Discussion by dpasukhi** [Open-Cascade-SAS/OCCT#31] [Announcements]
<div>
<p>Dear OCCT Community,</p>

<p>We are excited to announce a significant update to the OpenCASCADE Technology (OCCT). Our development team has introduced new functionality for STEP Import - User Defined Attributes (UDA, Metadata, property), which is now available in the latest master, <a href="https://git.dev.opencascade.org/gitweb/?p=occt.git;a=shortlog;h=refs/heads/CR0-WEEK-7">CR0-WEEK-7</a>.</p>

<p>The official recommended practices for STEP UDA can be found <a href="https://www.mbx-if.org/home/wp-content/uploads/2024/05/rec_prac_user_def_attributes_v18.pdf">here</a>.</p>

<p>This update addresses various issues reported by the community. One notable bugfix is detailed in the OCCT bugtracker issue: <a href="https://tracker.dev.opencascade.org/view.php?id=33530">33530</a>.</p>

<p>Below, you can see the result of importing some STEP files:</p>

<div><img alt="STEP Import Result 1" src="https://dev.opencascade.org/system/files/forum/nist_ftc_06.png" /></div>

<div><img alt="STEP Import Result 2" src="https://dev.opencascade.org/system/files/forum/nist_ctc_04_0.png" /></div>

<p>Imported metadata is available only in XCAF reading. Metadata is stored as a <i>TDataStd_NamedData(key-values)</i>&nbsp;attribute attached to the Shape label.</p>

<p>This enhancement opens up several possibilities for users:</p>

<ul>
	<li><strong>Custom Data Annotation:</strong> Users can now annotate STEP files with custom metadata, facilitating better organization and documentation of CAD models.</li>
	<li><strong>Workflow Automation:</strong> Automated processing of STEP files becomes more versatile with the ability to import, streamlining various engineering workflows.</li>
	<li><strong>Interoperability:</strong> Enhanced support for STEP metadata improves interoperability with other CAD software that utilize STEP format, fostering collaboration among different engineering teams.</li>
</ul>

<p>To dump available metadata in DRAWEXE, you can use the following command (the code located in <i>src\XDEDRAW\XDEDRAW_Shapes.cxx</i> <i>XGetProperties</i>):</p>

<pre>
<code># Prints named properties assigned to all document's shape labels or chosen labels of shapes
XGetProperties Doc [label1, label2, ...] [shape1, shape2, ...]</code></pre>

<p>To extract imported metadata by C++ code, you can use the following sample:</p>

<pre>
<code>Handle(TDocStd_Document) aDoc = ...;
Handle(XCAFDoc_ShapeTool) aShapeTool = XCAFDoc_DocumentTool::ShapeTool(aDoc-&gt;Main());
NCollection_IndexedDataMap&lt;TCollection_AsciiString, Handle(TDataStd_NamedData)&gt; aNameDataMap; // Container to keep label's entry and its metadata
for (TDF_ChildIterator anIter(aShapeTool-&gt;Label(), Standard_True); anIter.More(); anIter.Next())
{
  const TDF_Label&amp; aLabel = anIter.Value();
  TCollection_AsciiString anEntry;
  TDF_Tool::Entry(aLabel, anEntry);
  Handle(TDataStd_NamedData) aNamedData = aShapeTool-&gt;GetNamedProperties(aLabel);
  if (!aNamedData.IsNull())
  {
    aNameDataMap.Add(anEntry, aNamedData);
  }
}</code></pre>

<p>Moreover, we are actively working on additional features for the upcoming OCCT 7.9 release, including:</p>

<ul>
	<li>Importing metadata from GLTF files</li>
	<li>Exporting metadata into STEP files</li>
</ul>

<p>We believe these enhancements will further empower our community to leverage OCCT for various engineering and design tasks. Your feedback and contributions are invaluable as we continue to improve OCCT and support the needs of our users.</p>

<p>Best regards,<br />
The OCCT Development Team.</p>
</div>

**Comment by VM4Dim** [comment]
> The official recommended practices for STEP UDA can be found [here](https://www.mbx-if.org/documents/rec_prac_user_def_attributes_v18.pdf).

404 Not Found

**Reply by dpasukhi** [reply]
Thank you. Fixed.
